### PR TITLE
IGNITE-15501 Improvements in ConfigurationExtension - support for listeners and internal schema extensions

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-15501

### CODE REVIEW

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtensionTest.java [hunk 1]
```diff
@@ -50,4 +56,62 @@ public void injectConfiguration(
...
         assertEquals(300, paramCfg.joinTimeout().value());
     }
+
+    /** Tests that notifications work on injected configuration instance. */
+    @Test
+    public void notifications() {
+        List<String> log = new ArrayList<>();
+
+        fieldCfg.listen(ctx -> {
+            log.add("update");
```
Comment (sashapolo) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtensionTest.java:66]: are listeners executed in the same thread?

**Review by sashapolo — APPROVED** [review]

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtensionTest.java (hunk 1 above)
Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtensionTest.java:66]: yes

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 2]
```diff
@@ -129,7 +131,7 @@ private static Object cfgValue(
         Class<?> schemaClass = Class.forName(type.getCanonicalName() + "Schema");
 
         // Internal configuration extensions are not yet supported. This will probably be changed in the future.
```
Comment (sanpwc) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:133]: Seems it's not true anymore.

Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:133]: You're right

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 3]
```diff
@@ -160,16 +167,31 @@ private static Object cfgValue(
...
+
+                        ConfigurationNotificationsUtil.notifyListeners(
+                            sr.getRoot(rootKey),
+                            copy.getRoot(rootKey),
+                            ((DynamicConfiguration<InnerNode, ?>)cfgRef.get()),
+                            ++storageRev,
+                            futures
+                        );
+
+                        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
```
Comment (sanpwc) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:180]: In order to emulate configuration registry correctly, we should have following guaranties:
- same notification order as in real-life configuration;
- same non-fully-processing-awaiting logic of inner change notifications; 
- same fully-processing-awaiting logic of cross change notifications;
- same exception handling logic for exceptions thrown from listeners;

Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:180]: Thank you!

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtensionTest.java (hunk 1 above)
Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtensionTest.java:66]: Not anymore, but hb rules still apply, collection doesn't have to be concurrent

**Review by SammyVimes — APPROVED** [review]

**Pull request merged**